File: assignment 7/python/programs/3.py
class CircularQueue:
    def __init__(self, size):
        self.size = size
        self.queue = [None] * size
        self.front = self.rear = -1

    def is_empty(self):
        return self.front == -1

    def is_full(self):
        return (self.rear + 1) % self.size == self.front

    def enqueue(self, item):
        if self.is_full():
            print("Queue is full. Cannot enqueue item.")
        else:
            if self.is_empty():
                self.front = self.rear = 0
            else:
                self.rear = (self.rear + 1) % self.size
            self.queue[self.rear] = item
            print(f"Enqueued: {item}")

    def dequeue(self):
        if self.is_empty():
            print("Queue is empty. Cannot dequeue.")
        else:
            item = self.queue[self.front]
            if self.front == self.rear:
                self.front = self.rear = -1
            else:
                self.front = (self.front + 1) % self.size
            print(f"Dequeued: {item}")
            return item

    def display(self):
        if self.is_empty():
            print("Queue is empty.")
        else:
            print("Queue elements:", end=" ")
            i = self.front
            while True:
                print(self.queue[i], end=" ")
                if i == self.rear:
                    break
                i = (i + 1) % self.size
            print()


# Example usage:
if __name__ == "__main__":
    size = 5
    queue = CircularQueue(size)

    queue.enqueue(10)
    queue.enqueue(20)
    queue.enqueue(30)
    queue.enqueue(40)
    queue.enqueue(50)
    queue.display()

    queue.dequeue()
    queue.dequeue()
    queue.display()

    queue.enqueue(60)
    queue.enqueue(70)
    queue.display()

File: assignment 7/c/programs/3.c
#include <stdio.h>
#include <stdlib.h>

struct node
{
    int data;
    struct node *next;
} *front, *rear, *temp;

void enq(int data);
void deq();
void display();

int main()
{
    int choice, data, n;
    front = rear = NULL;

    printf("\n1. Enqueue\n2. Dequeue\n3. Display\n4. Exit\n");

    while (1)
    {
        printf("\nEnter your choice: ");
        scanf("%d", &choice);

        switch (choice)
        {
        case 1:
            printf("Enter the number of elements to be enqueued: ");
            scanf("%d", &n);
            for (int i = 0; i < n; i++)
            {
                printf("Enter data to be enqueued: ");
                scanf("%d", &data);
                enq(data);
            }
            break;
        case 2:
            printf("Enter the number of elements to be dequeued: ");
            scanf("%d", &n);
            for (int i = 0; i < n; i++)
            {
                deq();
            }
            break;
        case 3:
            display();
            break;
        case 4:
            printf("Exiting from app.\n");
            exit(0);
        default:
            printf("Invalid choice!\n");
        }
    }

    return 0;
}

void enq(int data)
{
    temp = (struct node *)malloc(sizeof(struct node));
    temp->data = data;
    if (front == NULL)
    {
        front = rear = temp;
        rear->next = front;
    }
    else
    {
        rear->next = temp;
        rear = temp;
        rear->next = front;
    }
}

void deq()
{
    if (front == NULL)
    {
        printf("Queue is empty!\n");
    }
    else if (front == rear)
    {
        printf("Dequeued element: %d\n", front->data);
        free(front);
        front = rear = NULL;
    }
    else
    {
        temp = front;
        printf("Dequeued element: %d\n", temp->data);
        front = front->next;
        rear->next = front;
        free(temp);
    }
}

void display()
{
    if (front == NULL)
    {
        printf("Queue is empty!\n");
    }
    else
    {
        temp = front;

        while (temp->next != front)
        {
            printf("%d ", temp->data);
            temp = temp->next;
        }

        printf("%d ", temp->data);
    }
}
